### PR TITLE
Add docs on decoupled final response feature

### PR DESCRIPTION
**Pull request by rmccorm4**

### CODE REVIEW

**Review by dyastremsky — DISMISSED** [review]

File: docs/user_guide/decoupled_models.md [hunk 1]
```diff
@@ -87,10 +87,100 @@ exactly one response per request. Even standard ModelInfer RPC in the GRPC endpo
...
 how the gRPC streaming can be used to infer decoupled models.
 
 If using [Triton's in-process C API](../customization_guide/inference_protocols.md#in-process-triton-server-api),
 your application should be cognizant that the callback function you registered with 
 `TRITONSERVER_InferenceRequestSetResponseCallback` can be invoked any number of times,
 each time with a new response. You can take a look at [grpc_server.cc](https://github.com/triton-inference-server/server/blob/main/src/grpc_server.cc)
+
+### Knowing When a Decoupled Inference Request is Complete
+
+A request is considered complete when a response containing the
```
Comment (nnshah1) [inline docs/user_guide/decoupled_models.md:100]: At a high level - I don't think we need a full description of the different possibilities here. Could we leave it as "a request is considered complete when a response containing the ... is received from a model backend". 

And then split into two sections, "Client applications using the C API can retrieve this via the ...", "Client applications using the streaming GRPC protocol can retrieve this using response parameters. Decoupled models may not send a response for each request but may send a response after receiving multiple requests. In these cases GRPC client can opt-in to receive an empty final response for each request. By default, empty final responses are not sent to save network traffic."

I believe the implementation for a backend should already be covered as that isn't changing - so the existing documentation on sending flags / setting the response complete - should be sufficient - and we shouldn't need to describe how a backend indicates the end - only how a client can know.

File: docs/user_guide/decoupled_models.md [hunk 2]
```diff
@@ -87,10 +87,100 @@ exactly one response per request. Even standard ModelInfer RPC in the GRPC endpo
...
+
+However, there are cases where a user may want to write a client that can
+generically handle any model, without knowing implementation details about it.
+Similarly, there are cases where the number of responses a model will send
+is unknown beforehand, so the client may need a programmatic way to know when
+the final response for a given request has been received. A common case for
+this may be where for a language model that has a `1` request -> `N` response
+structure. 
+
+To handle this case, Triton exposes a boolean `"triton_final_response"` response
```
Comment (nnshah1) [inline docs/user_guide/decoupled_models.md:142]: need to add to the extension_parameters doc.

Also - maybe a line on the sequence extension doc (we reference grpc streaming mode there).

I think streaming mode should probably be its own extension doc - and ideally the new documentation would actually live there - as the changes are only for GRPC streaming mode - no changes for C API or for backend in actuality -

File: docs/user_guide/decoupled_models.md [hunk 3]
```diff
@@ -87,10 +87,100 @@ exactly one response per request. Even standard ModelInfer RPC in the GRPC endpo
...
+#### TRITONBACKEND_ResponseFactorySendFlags 
+
+When a final response is sent via
+`TRITONBACKEND_ResponseFactorySendFlags(factory, TRITONSERVER_RESPONSE_COMPLETE_FINAL)`,
+no response is sent to the frontend by the backend/model, so nothing will be sent
+back to the client by default. For the client to receive a response containing
+this "final" signal via the `"triton_final_response"` parameter, the client
+will have to opt-in through the client library.
+
+To opt-in through the Python client library, the `enable_empty_final_response` arg
```
Comment (Tabrizian) [inline docs/user_guide/decoupled_models.md:168]: Need to mention that this is only available for gRPC.

**Review by Tabrizian — COMMENTED** [review]
@rmccorm4 would be great if we can finalize this PR by EOD so that I can include a link to this in the release notes.

File: docs/user_guide/decoupled_models.md (hunk 2 above)
Comment (rmccorm4) [reply docs/user_guide/decoupled_models.md]: 1. Added mention of the reserved params in parameters doc.
2. Took a look at sequence doc, but not sure where it makes sense to add anything at the moment.
3. Agreed streaming extension could be its own doc, but will defer to a future PR. Maybe along with KServe discussions.

Lmk what you think.

File: docs/user_guide/decoupled_models.md (hunk 1 above)
Comment (rmccorm4) [reply docs/user_guide/decoupled_models.md]: done

File: docs/user_guide/decoupled_models.md (hunk 3 above)
Comment (rmccorm4) [reply docs/user_guide/decoupled_models.md]: done

**Review by Tabrizian — DISMISSED** [review]

File: docs/user_guide/decoupled_models.md (hunk 2 above)
Comment (nnshah1) [reply docs/user_guide/decoupled_models.md]: Sounds good

**Review by nnshah1 — APPROVED** [review]

**Comment by rmccorm4** [comment]
Noticed small [typo](https://github.com/triton-inference-server/server/pull/5936/commits/cee930e30ef88930a599999c9cc2f9cacc328e9d) at the end, should be good now.

**Review by Tabrizian — APPROVED** [review]
We need to cherry-pick it to 23.06 too.

**Pull request merged**